src/dataflow/analysis/analysis.ts

```diff
@@ -10,67 +10,229 @@
+    // Remove unresolved parent, and replace with inherited unresolved parents.
+    const modifierSet = this.unresolvedFlows.get(parentExpr.edge);
+    this.unresolvedFlows.delete(parentExpr.edge);
 
-  const finalResult = new ValidationResult();
+    for (const modifier of modifierSet) {
+      // Copy flows from parent (and apply modifiers).
+      const newFlows = parentExpr.resolvedFlows.copyAndModify(modifier);
+      this.resolvedFlows.addAll(newFlows);
 
-  // Check every input path into the given edge.
-  // NOTE: This is very inefficient. We check every single check condition against every single edge in every single input path.
-  for (const path of allInputPaths(edgeToCheck)) {
-    const flow = computeFlowForPath(path);
-    if (!flow.evaluateCheck(check)) {
-      finalResult.failures.push(`'${check.originalCheck.toManifestString()}' failed for path: ${path.toString()}`);
+      // Copy any unresolved grandparents (and apply modifiers).
+      for (const [edge, parentModifierSet] of parentExpr.unresolvedFlows) {
+        for (const parentModifier of parentModifierSet) {
+          // Combine modifiers, applying parent's modifiers first.
+          const combinedModifier = parentModifier.copyAndModify(modifier);
+          this.inheritFromEdge(edge, combinedModifier);
+        }
+      }
     }
   }
 
-  return finalResult;
+  /** Add a new unresolved flow, consisting of the given edge and a modifier for it. */
+  private inheritFromEdge(edge: Edge, modifier: FlowModifier) {
+    if (this.unresolvedFlows.has(edge)) {
+      this.unresolvedFlows.get(edge).add(modifier);
+    } else {
+      this.unresolvedFlows.set(edge, new FlowModifierSet(modifier));
+    }
+  }
+
+  toString() {
+    const result: string[] = [`EdgeExpression(${this.edge.label}) {`];
+
+    for (const flow of this.resolvedFlows) {
+      result.push('  ' + flow.toUniqueString());
+    }
+    for (const [edge, modifierSets] of this.unresolvedFlows) {
+      for (const modifiers of modifierSets) {
+        result.push(`  ${edge.label} + ${modifiers.toString()}`);
+      }
+    }
+    result.push('}');
+    return result.join('\n');
+  }
 }
 
-/**
- * Collects all the tags, nodes and edges along the given path, canceling tag claims that are
- * negated by "not" claims for the same tag downstream in the path, and ignoring
- * "not" claims without corresponding positive claims upstream, as these dangling
- * "not" claims are irrelevant for the given path. Note that "derives from" claims
- * only prune paths, and are dealt with during path generation, so they are ignored.
- */
-function computeFlowForPath(path: BackwardsPath): Flow {
-  const flow = new Flow();
-  for (const edge of path.edgesInForwardDirection()) {
-    if (edge.modifier) {
-      flow.modify(edge.modifier);
+class Solver {
+  readonly graph: FlowGraph;
+
+  /** Maps from an edge to a "expression" for it. */
+  readonly edgeExpressions: Map<Edge, EdgeExpression> = new Map();
```

Comment:
The EdgeExpressions for each parent edge are not necessarily known at the time the EdgeExpression is constructed. We keep a reference to each parent Edge, and later on resolve those Edges into their EdgeExpressions when we know what they are.

The only other thing this could be is a Map<string, FlowModifierSet>, with Edge IDs as its string key. I didn't think that was any clearer though.

It could be an array instead of a map, but then lookups and deletions become more expensive.